packages/thirdweb/src/extensions/marketplace/offers/offers.test.ts

```diff
@@ -0,0 +1,346 @@
+
+describe("Marketplace: Offers", () => {
+  let nftTokenId: bigint;
+  let marketplaceContract: ThirdwebContract;
+  let erc721Contract: ThirdwebContract;
+  let WETH9: ThirdwebContract;
+  beforeAll(async () => {
+    const marketplaceAddress = await deployMarketplaceContract({
+      account: TEST_ACCOUNT_A,
+      chain: ANVIL_CHAIN,
+      client: TEST_CLIENT,
+      params: {
+        name: "TestMarketPlace",
+      },
+    });
+    marketplaceContract = getContract({
+      address: marketplaceAddress,
+      client: TEST_CLIENT,
+      chain: ANVIL_CHAIN,
+    });
+
+    // also deploy an ERC721 contract
+    const erc721Address = await deployERC721Contract({
+      type: "TokenERC721",
+      account: TEST_ACCOUNT_A,
+      chain: ANVIL_CHAIN,
+      client: TEST_CLIENT,
+      params: {
+        name: "TestERC721",
+      },
+    });
+
+    erc721Contract = getContract({
+      address: erc721Address,
+      client: TEST_CLIENT,
+      chain: ANVIL_CHAIN,
+    });
+
+    const mintTransaction = mintTo({
+      contract: erc721Contract,
+      to: TEST_ACCOUNT_A.address,
+      nft: { name: "Test:ERC721:Offers" },
+    });
+    const receipt = await sendAndConfirmTransaction({
+      transaction: mintTransaction,
+      account: TEST_ACCOUNT_A,
+    });
+
+    const mintEvents = parseEventLogs({
+      events: [tokensMintedEvent()],
+      logs: receipt.logs,
+    });
+
+    expect(mintEvents.length).toBe(1);
+    expect(mintEvents[0]?.args.tokenIdMinted).toBeDefined();
+
+    nftTokenId = mintEvents[0]?.args.tokenIdMinted as bigint;
+
+    // biome-ignore lint/style/noNonNullAssertion: test only and we know it will exist here because it gets deployed in the beforeAll
+    WETH9 = (await getDeployedInfraContract({
+      chain: ANVIL_CHAIN,
+      contractId: "WETH9",
+      constructorParams: [],
+      client: TEST_CLIENT,
+    }))!;
+
+    expect(WETH9.address).toMatchInlineSnapshot(
+      `"0x81e609b897393731a3d23c1d311330340cebb9e9"`,
+    );
+    // does a lot of stuff, this may take a while
+  }, 120_000);
+
+  // NOTE: this test suite relies on the order of operations in order to not re-deploy contracts over and over
+  // this is generally not ideal, but we're doing it here to save time
+
+  it("should start out with 0 offers", async () => {
+    await expect(totalOffers({ contract: marketplaceContract })).resolves.toBe(
+      0n,
+    );
+    await expect(
+      getAllOffers({ contract: marketplaceContract }),
+    ).resolves.toMatchInlineSnapshot("[]");
+    await expect(
+      getAllValidOffers({ contract: marketplaceContract }),
+    ).resolves.toMatchInlineSnapshot("[]");
+  });
+
+  it("should allow a user to create an offer", async () => {
+    // first deposit some tokens into the WETH9 contract
+    // do do that, determine the address of the WETH9 contract
+
+    await expect(
+      sendAndConfirmTransaction({
+        transaction: deposit({ contract: WETH9, amount: "10" }),
```

Comment:
Still weird to.me.that we have to do this. How did we handle it in V4?

Reply:
like this! we assume you already have the tokens and then we approve for u